components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/internal/impl/DiscoverableApplicationManagerImpl.java

```diff
@@ -34,8 +36,20 @@ public List<ApplicationBasicInfo> getDiscoverableApplicationBasicInfo(int limit,
             filter, String sortOrder, String sortBy, String tenantDomain) throws
             IdentityApplicationManagementException {
 
-        ApplicationDAO appDAO = ApplicationMgtSystemConfig.getInstance().getApplicationDAO();
-        return appDAO.getDiscoverableApplicationBasicInfo(limit, offset, filter, sortOrder, sortBy, tenantDomain);
+        try {
+            ApplicationDAO appDAO = ApplicationMgtSystemConfig.getInstance().getApplicationDAO();
+            if (ApplicationMgtUtil.isSubOrg(tenantDomain)) {
+                String primaryOrgID = ApplicationMgtUtil.getParentOrgId(tenantDomain);
+                return appDAO.getDiscoverableAppsInfoFromRootAndSubOrg(limit, offset, filter, sortOrder, sortBy,
+                        tenantDomain, primaryOrgID);
+            } else {
+                return appDAO.getDiscoverableApplicationBasicInfo(limit, offset, filter, sortOrder, sortBy,
+                        tenantDomain);
+            }
+        } catch (OrganizationManagementException e) {
+            throw new IdentityApplicationManagementException("Error while checking whether the tenant domain is a " +
+                    "sub-organization.", e);
```

Comment:
Can't we do something similar ?
```suggestion
            throw new IdentityApplicationManagementException(e.getErrorMessage(), e);
```